File: phase-1/refactoring/php/src/AssertionActionService.php
<?php

namespace App;

use Monolog\Logger;
use Monolog\Handler\StreamHandler;

class AssertionActionService implements IAssertionActionService
{
    private $logger;

    public function __construct()
    {
        $this->logger = new Logger('assertionLogger');
        $this->logger->pushHandler(new StreamHandler('path/to/your/logs.log', Logger::WARNING));
    }

    public function getAssertionActionResponse($assertionConsumerServiceRequest, $claimsInfos, $requestedEmail)
    {
        if ($assertionConsumerServiceRequest->isClaimsRequired && !$this->validateMandatoryClaims($claimsInfos, $requestedEmail)) {
            return [
                'status' => 'Error',
                'message' => 'The mandatory claims are not correctly configured in the identity provider'
            ];
        }

        return [
            'status' => 'Success'
        ];
    }

    private function validateMandatoryClaims($claimsInfos, $requestedEmail)
    {
        $claimsIsValid = true;

        if (empty($claimsInfos->email)) {
            $this->logger->warning("The claim Email is not correctly configured in the identity provider for this user {$requestedEmail}");
            $claimsIsValid = false;
        }

        if (empty($claimsInfos->userName)) {
            $this->logger->warning("The claim UserName is not correctly configured in the identity provider for this user {$requestedEmail}");
            $claimsIsValid = false;
        }

        if (empty($claimsInfos->firstName)) {
            $this->logger->warning("The claim FirstName is not correctly configured in the identity provider for this user {$requestedEmail}");
            $claimsIsValid = false;
        }

        if (empty($claimsInfos->lastName)) {
            $this->logger->warning("The claim LastName is not correctly configured in the identity provider for this user {$requestedEmail}");
            $claimsIsValid = false;
        }

        return $claimsIsValid;
    }
}

File: phase-1/refactoring/elixir/lib/assertion_action_service.ex
defmodule Refactoring.AssertionActionService do
  require Logger

  def get_assertion_action_response(%{is_claims_required: true} = request, claims_infos, requested_email) do
    if request.is_claims_required and not validate_mandatory_claims(claims_infos, requested_email) do
      %{
        status: :error,
        message: "The mandatory claims are not correctly configured in the identity provider"
      }
    else
      %{status: :success}
    end
  end

  def get_assertion_action_response(_request, _claims_infos, _requested_email) do
    %{status: :success}
  end

  
  defp validate_mandatory_claims(claims_infos, requested_email) do

    claims_is_valid = if claims_infos.email == nil or claims_infos.email == "" do
      Logger.warning("The claim email is not correctly configured in the identity provider for this user #{requested_email}", [])
      false
    else
      true
    end

    claims_is_valid = if claims_infos.username == nil or claims_infos.username == "" do
      Logger.warning("The claim username is not correctly configured in the identity provider for this user #{requested_email}", [])
      false
    else
      claims_is_valid
    end

    claims_is_valid = if claims_infos.first_name == nil or claims_infos.first_name == "" do
      Logger.warning("The claim first_name is not correctly configured in the identity provider for this user #{requested_email}", [])
      false
    else
      claims_is_valid
    end

    claims_is_valid = if claims_infos.last_name == nil or claims_infos.last_name == "" do
      Logger.warning("The claim last_name is not correctly configured in the identity provider for this user #{requested_email}", [])
      false
    else
      claims_is_valid
    end

    claims_is_valid
  end
end
